/**
 * Generated by [[bio.ferlab.datalake.spark3.utils.ClassGenerator]]
 * on 2023-12-06T19:41:53.428400
 */
package bio.ferlab.clin.etl.model.raw


case class RawFranklin(`variants`: Seq[VARIANTS] = Seq(VARIANTS()),
                       `batch_id`: String = "1",
                       `family_id`: String = "1",
                       `aliquot_id`: String = "12345",
                       `analysis_id`: String = "20")

case class PREDICTIONS(`ada`: Option[Double] = Some(0.5), // Using a non None value to avoid casting issues in tests
                       `aggregated_predictions`: Double = 0.36948757763975154,
                       `dbsnp`: String = "rs35354874",
                       `fathmm`: Double = -0.16,
                       `fitcons`: Double = 0.67177,
                       `genocanyon`: Double = 0.0209643034275722,
                       `gerp`: Double = 4.01,
                       `ma`: Double = 2.045,
                       `metalr`: Double = 0.2029,
                       `mt`: Double = 0.003803000000000001,
                       `revel`: Double = 0.257,
                       `sift`: Double = 0.008,
                       `splice_ai`: Double = 0.0)

case class ANNOTATIONS(`clinical_evidences`: CLINICAL_EVIDENCES = CLINICAL_EVIDENCES(),
                       `frequencies`: FREQUENCIES = FREQUENCIES(),
                       `predictions`: PREDICTIONS = PREDICTIONS(),
                       `transcripts`: Seq[TRANSCRIPTS] = Seq(TRANSCRIPTS()))

case class TRANSCRIPTS(`closest_distance_to_exon`: Long = -226,
                       `closest_exon`: Long = 2,
                       `effect`: String = "NON_SYNONYMOUS",
                       `exon_number`: Long = 2,
                       `gene`: String = "MUC3A",
                       `hgvc_c`: String = "c.8641C>A",
                       `hgvc_p`: String = "p.Leu2881Ile",
                       `region`: String = "EXONIC",
                       `transcript`: String = "NM_005960.2",
                       `transcript_exon_count`: Long = 12,
                       `transcript_type`: String = "REFSEQ",
                       `transcripts_count`: Long = 1)

case class CLINVAR(`submissions_by_classification`: Seq[SUBMISSIONS_BY_CLASSIFICATION] = Seq(SUBMISSIONS_BY_CLASSIFICATION()))

case class VARIANT(`alt`: String = "A",
                   `chromosome`: String = "chr7",
                   `end_position`: Long = 100960420,
                   `ref`: String = "C",
                   `start_position`: Long = 100960420,
                   `variation_type`: String = "SNP")

case class ACMG_RULES(`influence`: String = "StrongPathogenic",
                      `is_met`: Boolean = false,
                      `name`: String = "PS1",
                      `weight`: Long = 7)

case class CLINICAL_EVIDENCES(`clinvar`: CLINVAR = CLINVAR(),
                              `uniprot`: UNIPROT = UNIPROT())

case class SUBMISSIONS_BY_CLASSIFICATION(`count`: Long = 2,
                                         `level`: String = "BENIGN")

case class UNIPROT(`uniprot_id`: String = "P15923",
                   `variant_link`: String = "https://www.uniprot.org/uniprot/P15923")

case class CLASSIFICATION(`acmg_classification`: String = "UNCERTAIN_SIGNIFICANCE",
                          `acmg_rules`: Seq[ACMG_RULES] = Seq(ACMG_RULES()))

case class FREQUENCIES(`aggregated_frequency`: Double = 0.018720927,
                       `max_gnomad_frequency`: Double = 0.06544256210327148)

case class SAMPLE_DATA(`confidence_level`: String = "LOW",
                       `depth`: Long = 359,
                       `depth_alt`: Long = 164,
                       `depth_ref`: Long = 195,
                       `gq`: Double = 47.0,
                       `mapping_quality`: Double = 248.8,
                       `pl`: String = "84:0:47",
                       `quality`: Double = 49.47,
                       `vaf`: Double = 0.4568245125348189,
                       `zygosity`: String = "HET")

case class VARIANTS(`annotations`: ANNOTATIONS = ANNOTATIONS(),
                    `classification`: CLASSIFICATION = CLASSIFICATION(),
                    `included_in_analysis_workbench`: Boolean = false,
                    `priority`: PRIORITY = PRIORITY(),
                    `sample_data`: SAMPLE_DATA = SAMPLE_DATA(),
                    `variant`: VARIANT = VARIANT(),
                    `variant_franklin_link`: String = "https://franklin.genoox.com/clinical-db/variant/snp/chr7-100960420-C-A-HG38")

case class PRIORITY(`score`: Double = 0.0036969461275872164)