/**
 * Generated by [[bio.ferlab.datalake.testutils.ClassGenerator]]
 * on 2024-02-01T14:05:36.071053
 */
package bio.ferlab.clin.etl.model.raw

import bio.ferlab.clin.etl.model.normalized.{EXON, INTRON}
import bio.ferlab.clin.etl.model.raw


case class VCF_SNV_Somatic_Input(`contigName`: String = "chr1",
                                 `start`: Long = 69896,
                                 `end`: Long = 69897,
                                 `names`: List[String] = List("rs200676709"),
                                 `referenceAllele`: String = "T",
                                 `alternateAlleles`: List[String] = List("C"),
                                 `qual`: Option[Double] = Some(2324.17),
                                 `INFO_FILTERS`: List[String] = List("PASS"),
                                 `splitFromMultiAllelic`: Boolean = false,
                                 `INFO_ReadPosRankSum`: Option[Double] = Some(0.736),
                                 `INFO_hotspot`: Boolean = false,
                                 `INFO_DP`: Int = 330,
                                 `INFO_AQ`: Option[List[Double]] = Some(List(67.22)),
                                 `INFO_MQ`: Double = 15.54,
                                 `INFO_CSQ`: List[INFO_CSQ_SOMATIC] = List(raw.INFO_CSQ_SOMATIC()),
                                 `INFO_DB`: Boolean = true,
                                 `INFO_MQRankSum`: Option[Double] = Some(2.054),
                                 `INFO_FractionInformativeReads`: Option[Double] = Some(0.808),
                                 `INFO_OLD_MULTIALLELIC`: Option[String] = None,
                                 `genotypes`: List[SNV_SOMATIC_GENOTYPES] = List(SNV_SOMATIC_GENOTYPES(), SNV_SOMATIC_GENOTYPES(`sampleId` = "22222", `calls` = List(0, 0)), SNV_SOMATIC_GENOTYPES(`sampleId` = "33333")),
                                 `normalizationStatus`: NORMALIZATIONSTATUS = NORMALIZATIONSTATUS())

case class INFO_CSQ_SOMATIC(`Allele`: String = "C",
                            `Consequence`: List[String] = List("synonymous_variant"),
                            `IMPACT`: String = "LOW",
                            `SYMBOL`: String = "OR4F5",
                            `Gene`: String = "ENSG00000186092",
                            `Feature_type`: String = "Transcript",
                            `Feature`: String = "ENST00000335137",
                            `BIOTYPE`: String = "protein_coding",
                            `EXON`: SOMATIC_EXON = SOMATIC_EXON("1", "1"),
                            `INTRON`: SOMATIC_INTRON = SOMATIC_INTRON(None, None),
                            `HGVSc`: String = "ENST00000335137.4:c.807T>C",
                            `HGVSp`: Option[String] = Some("ENSP00000334393.3:p.Ser269%3D"),
                            `cDNA_position`: String = "843",
                            `CDS_position`: Option[String] = Some("807"),
                            `Protein_position`: Option[String] = Some("269"),
                            `Amino_acids`: AMINO_ACIDS = AMINO_ACIDS(Some("S"), None),
                            `Codons`: CODONS = CODONS(Some("tcT"), Some("tcC")),
                            `Existing_variation`: List[String] = List("rs200676709"),
                            `DISTANCE`: Option[String] = None,
                            `STRAND`: String = "1",
                            `FLAGS`: Option[List[String]] = None,
                            `PICK`: String = "1",
                            `VARIANT_CLASS`: String = "SNV",
                            `SYMBOL_SOURCE`: String = "HGNC",
                            `HGNC_ID`: String = "HGNC:14825",
                            `CANONICAL`: String = "YES",
                            `RefSeq`: Option[String] = Some("NM_001005484.1&NM_001005484.1&NM_001005484.2"),
                            `REFSEQ_MATCH`: Option[String] = None,
                            `SOURCE`: String = "Ensembl",
                            `REFSEQ_OFFSET`: Option[String] = None,
                            `HGVS_OFFSET`: Option[String] = None,
                            `HGVSg`: String = "chr1:g.69897T>C",
                            `CLIN_SIG`: Option[String] = None,
                            `SOMATIC`: Option[String] = None,
                            `PHENO`: Option[String] = None,
                            `PUBMED`: Option[String] = Some("29135816"))

case class SNV_SOMATIC_GENOTYPES(`sampleId`: String = "11111",
                                 `F1R2`: List[Int] = List(0, 0),
                                 `SQ`: List[Double] = List(56.08),
                                 `SB`: List[Int] = List(0, 0, 191, 141),
                                 `F2R1`: List[Int] = List(0, 166),
                                 `alleleDepths`: List[Int] = List(0, 30),
                                 `AF`: List[Double] = List(1.0),
                                 `phased`: Boolean = false,
                                 `calls`: Seq[Int] = Seq(0, 1),
                                 `depth`: Int = 1,
                                 `PS`: Option[Int] = None,
                                 `MB`: Option[List[Int]] = None)

case class AMINO_ACIDS(`reference`: Option[String] = None,
                       `variant`: Option[String] = None)

case class NORMALIZATIONSTATUS(`changed`: Boolean = false,
                               `errorMessage`: Option[String] = None)

case class CODONS(`reference`: Option[String] = None,
                  `variant`: Option[String] = None)

case class SOMATIC_EXON(`rank`: String = "1",
                        `total`: String = "1")

case class SOMATIC_INTRON(`rank`: Option[String] = None,
                          `total`: Option[String] = None)
